src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs

```diff
@@ -2401,6 +2402,56 @@ private bool getObjectContent(CORINFO_OBJECT_STRUCT_* objPtr, byte* buffer, int
             return ObjectToHandle(HandleToObject(objPtr).ObjectType);
         }
 
+        private CORINFO_METHOD_STRUCT_* getMethodFromDelegate(CORINFO_CLASS_STRUCT_* calledCls, CORINFO_OBJECT_STRUCT_* delegateObj, CORINFO_CLASS_STRUCT_** methodCls, CORINFO_CLASS_STRUCT_** targetCls)
+        {
+            Debug.Assert(calledCls != null);
+            Debug.Assert(delegateObj != null);
+
+            FrozenObjectNode frozenObject = HandleToObject(delegateObj);
+            MethodDesc method = frozenObject.DelegateMethod;
+
+            TypeDesc calledType = HandleToObject(calledCls);
+            if (!frozenObject.ObjectType.CanCastTo(calledType))
+            {
+                return null;
+            }
+
+            if (methodCls != null)
+            {
+                *methodCls = ObjectToHandle(method.OwningType);
+            }
+
+            if (targetCls != null)
+            {
+                DelegateInfo delegateInfo = _compilation.TypeSystemContext.GetDelegateInfo(calledType);
+                int paramCountDelegateClosed = delegateInfo.Signature.Length + 1;
+                int paramCountTargetMethod = method.Signature.Length;
+                if (!method.Signature.IsStatic)
+                {
+                    paramCountTargetMethod++;
+                }
+
+                CORINFO_CLASS_STRUCT_* target = null;
+                if (paramCountDelegateClosed != paramCountTargetMethod)
+                {
+                    TypeDesc targetType = frozenObject.DelegateTargetType;
+                    if (targetType == null)
+                    {
+                        // avoid delegates closed over null
+                        Debug.Assert(method.Signature.IsStatic);
```

Comment:
Can I safely rely on NativeAOT never seeing delegates closed over null here? Since it seems to use a null target for all static calls.

Reply:
I think that matches this assertion. Might want to move it out of the `if` and keep the same shape (it's either null and static method, or non-null and instance method):

https://github.com/dotnet/runtime/blob/0d62887a30553b8177dc90f9e39559be0e6c7707/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs#L2802